pyrokinetics/pyro.py

```diff
@@ -1,339 +1,1423 @@
+        """
+        try:
+            if self.kinetics is None:
+                raise AttributeError
+        except AttributeError:
+            raise RuntimeError(
+                "Pyro.load_local_species: Must have read global kinetics first. "
+                "Use function load_global_kinetics."
+            )
 
-        self._float_format = value
+        if psi_n < 0 or psi_n > 1:
+            raise ValueError(
+                "Pyro.load_local_species: psi_n must be between 0 and 1, received "
+                f"{psi_n}."
+            )
 
-    def __deepcopy__(self, memodict):
+        if a_minor is None:
+            try:
+                if self.eq is None:
+                    raise AttributeError
+                a_minor = self.eq.a_minor
+            except AttributeError:
+                raise RuntimeError(
+                    "Pyro.load_local_species: Must set a_minor, or read global "
+                    "equilibrium first. To set global_equilibrium, use function "
+                    "load_global_equilibrium."
+                )
+
+        local_species = LocalSpecies()
+        local_species.from_kinetics(self.kinetics, psi_n=psi_n, lref=a_minor)
+        self.local_species = local_species
+
+    def load_local(self, psi_n: float, local_geometry: str = "Miller") -> None:
         """
-        Allows for deepcopy of a Pyro object
+        Combines calls to ``load_local_geometry()`` and ``load_local_species()``
+
+        Parameters
+        ----------
+        psi_n: float
+            Normalised flux surface on which to calculate local geometry. 0 is the
+            center of the equilibrium, 1 is the Last-Closed-Flux-Surface (LCFS).
+        local_geometry: str, default "Miller"
+            The type of LocalGeometry to create, expressed as a string. Must be in
+            ``supported_local_geometries``.
 
         Returns
         -------
-        Copy of pyro object
+        ``None``
+
+        Raises
+        ------
+        Exception
+            See exceptions for ``load_local_geometry()`` and ``load_local_species()``.
         """
-        from copy import deepcopy
+        self.load_local_geometry(psi_n, local_geometry=local_geometry)
+        self.load_local_species(psi_n)
+
+    # Utility for copying Pyro object
 
+    def __deepcopy__(self, memodict):
+        """
+        Create a new Pyro, recursively copying all structures.
+
+        Returns
+        -------
+        Pyro
+            Deep copy of self.
+        """
         new_pyro = Pyro()
 
         for key, value in self.__dict__.items():
-            setattr(new_pyro, key, deepcopy(value))
+            setattr(new_pyro, key, copy.deepcopy(value))
 
         return new_pyro
+
+    # Add properties that allow direct access to GKInput.data
+    # TODO This feels dangerous... could use a refactor
+    # TODO Not sure how to automate generation of these when new gk_codes are added
+    @property
+    def gs2_input(self) -> Union[f90nml.Namelist, None]:
```

Comment:
I think this is probably fine as a stop-gap for now, but we do really need a better overall design, I feel.

This might mean, for instance, that the code-specific portions live completely separately from the code-independent part, and the API looks more like, for example:


```python
gs2 = GS2(input_file)  # Read in, no conversion
pyro = gs2.to_pyro()  # Normalise to canonical form
CGYRO.write(pyro, cgyro_template)  # Convert and write to CGYRO
GENE.write(pyro, gene_template, dict_of_extras)
```

Obviously this would be a significant change to user code, but would hopefully end up with a lot cleaner, clearer, and more extensible library